packages/commons/src/auth/authorization.ts

```diff
@@ -0,0 +1,210 @@
+import { unauthorizedError } from "pagopa-interop-models";
+import { P, match } from "ts-pattern";
+import { AppContext } from "../context/context.js";
+import { NonEmptyArray } from "../index.js";
+import {
+  UserRole,
+  UIAuthData,
+  M2MAuthData,
+  InternalAuthData,
+  MaintenanceAuthData,
+  AuthData,
+  SystemRole,
+} from "./authData.js";
+
+type AuthRole = UserRole | SystemRole;
+
+type ContainsAuthRole<
+  Arr extends AuthRole[],
+  V extends AuthRole
+> = Arr extends [infer Head, ...infer Tail extends AuthRole[]]
+  ? Head extends V
+    ? true
+    : ContainsAuthRole<Tail, V>
+  : false;
+
+type ExtractUserRoles<Arr extends NonEmptyArray<AuthRole>> = Exclude<
+  Arr[number],
+  SystemRole
+>;
+
+type AllowedAuthData<AdmittedRoles extends NonEmptyArray<AuthRole>> =
+  // If "m2m" is in the array, add M2MAuthData
+  | ([ContainsAuthRole<AdmittedRoles, "m2m">] extends [true]
+      ? M2MAuthData
+      : never)
+  // If "internal" is in the array, add InternalAuthData
+  | ([ContainsAuthRole<AdmittedRoles, "internal">] extends [true]
+      ? InternalAuthData
+      : never)
+  // If "maintenance" is in the array, add MaintenanceAuthData
+  | ([ContainsAuthRole<AdmittedRoles, "maintenance">] extends [true]
+      ? MaintenanceAuthData
+      : never)
+  // If there are user roles in the array, add UIAuthData
+  | ([ExtractUserRoles<AdmittedRoles>] extends [never] ? never : UIAuthData);
+
+export function validateAuthorization<
+  AdmittedRoles extends NonEmptyArray<AuthRole>
+>(
+  ctx: AppContext,
+  admittedAuthRoles: AdmittedRoles
+): asserts ctx is AppContext<AllowedAuthData<AdmittedRoles>> {
+  const { authData } = ctx;
+
+  match(authData)
+    .with(
+      { systemRole: P.union("m2m", "internal", "maintenance") },
```

Comment:
If we define `systemRole`, in case there will be new roles in the future, this could be
```suggestion
      { systemRole: P.union(...Object.values(SystemRole)) },
```

Reply:
It's exhaustive so we are still safe in the future: TS will force us to handle new roles here.
I prefer keeping pattern matching lines explicit and readable, using literals and enforcing safety by using .exhaustive

Reply:
> It's exhaustive so we are still safe in the future: TS will force us to handle new roles here. I prefer keeping pattern matching lines explicit and readable, using literals and enforcing safety by using .exhaustive

You're right but if we don't want to use the `Object.values` I suggest to use their defined const

Reply:
Done 👍